tensorflow_model_optimization/python/core/clustering/keras/cluster.py

```diff
@@ -0,0 +1,213 @@
+  keras.models.save_model(clustered_model, keras_file)
+
+  with clustering_scope():
+    loaded_model = keras.models.load_model(keras_file)
+  ```
+  """
+  return custom_object_scope(
+    {
+      'ClusterWeights': ClusterWeights,
+    }
+  )
+
+
+def cluster_weights(to_cluster, number_of_clusters, cluster_centroids_init, **kwargs):
+  """Modify a keras layer or model to be clustered during training.
+
+  This function wraps a keras model or layer with clustering functionality
+  which clusters the layer's weights during training. For examples, using
+  this with number of clusters equals 8 will ensure that each weight tensor has
+  no more than 8 unique values.
+
+  Before passing to the clustering API, a model should already be trained and show some acceptable performance
+  on the testing/validation sets.
+
+  The function accepts either a single keras layer
+  (subclass of `keras.layers.Layer`), list of keras layers or a keras model
+  (instance of `keras.models.Model`) and handles them appropriately.
+
+  If it encounters a layer it does not know how to handle, it will throw an
+  error. While clustering an entire model, even a single unknown layer would lead
+  to an error.
+
+  Cluster a model:
+
+  ```python
+  clustering_params = {
+    'number_of_clusters': 8,
+    'cluster_centroids_init': 'density-based'
+  }
+
+  clustered_model = cluster_weights(original_model, **clustering_params)
+  ```
+
+  Cluster a layer:
+
+  ```python
+  clustering_params = {
+    'number_of_clusters': 8,
+    'cluster_centroids_init': 'density-based'
+  }
+
+  model = keras.Sequential([
+      layers.Dense(10, activation='relu', input_shape=(100,)),
+      cluster_weights(layers.Dense(2, activation='tanh'), **clustering_params)
+  ])
+  ```
+
+  Arguments:
+      to_cluster: A single keras layer, list of keras layers, or a
+        `tf.keras.Model` instance.
+      number_of_clusters: the number of cluster centroids to form when clustering a layer/model.
+        For example, if number_of_clusters=8 then only 8 unique values will be used in each weight array.
+      cluster_centroids_init: how to initialize the cluster centroids.
+        Can have following values:
+          1. 'random' : centroids are sampled using the uniform distribution between the minimum and maximum weight
+          values in a given layer
+          2. 'density-based' : density-based sampling. First, cumulative distribution function is built for weights,
+          then y-axis is evenly spaced into number_of_clusters regions. After this the corresponding x values are
+          obtained and used to initialize clusters centroids.
+          3. 'linear' : cluster centroids are evenly spaced between the minimum and maximum values of a given weight
+      **kwargs: Additional keyword arguments to be passed to the keras layer.
+        Ignored when to_cluster is not a keras layer.
+
+  Returns:
+    Layer or model modified with clustering wrappers.
```

Comment:
No change needed: more of a note for us in the future, for pruning also. 

It's weird for these docs to make a reference to "clustering wrappers", which is an implementation detail that users don't need to know about.

Reply:
I rephrased this bit. Have a look a let me know if it sounds better to you.